File: app/src/main/java/com/chefvoice/app/voice/IngredientParser.kt
package com.chefvoice.app.voice

import com.chefvoice.app.model.Ingredient

/**
 * Parses an ingredient phrase after speech recognition or manual entry.
 *
 * The parser intentionally normalizes common ASR spellings such as
 * "tea spoons" -> "teaspoons" and "to teaspoons" -> "2 teaspoons".
 */
object IngredientParser {
    private val unitAliases = linkedMapOf(
        "tablespoonfuls" to "tbsp", "tablespoonful" to "tbsp", "tablespoons" to "tbsp", "tablespoon" to "tbsp", "tbsp" to "tbsp", "tbs" to "tbsp",
        "teaspoonfuls" to "tsp", "teaspoonful" to "tsp", "teaspoons" to "tsp", "teaspoon" to "tsp", "tsp" to "tsp",
        "kilograms" to "kg", "kilogram" to "kg", "kg" to "kg",
        "milligrams" to "mg", "milligram" to "mg", "mg" to "mg",
        "grams" to "g", "gram" to "g", "g" to "g",
        "liters" to "L", "liter" to "L", "litres" to "L", "litre" to "L", "l" to "L",
        "milliliters" to "ml", "milliliter" to "ml", "millilitres" to "ml", "millilitre" to "ml", "ml" to "ml",
        "fluid ounces" to "fl oz", "fluid ounce" to "fl oz", "fl oz" to "fl oz",
        "ounces" to "oz", "ounce" to "oz", "oz" to "oz",
        "pounds" to "lb", "pound" to "lb", "lbs" to "lb", "lb" to "lb",
        "cupfuls" to "cup", "cupful" to "cup", "cups" to "cup", "cup" to "cup",
        "chunks" to "chunk", "chunk" to "chunk",
        "cloves" to "clove", "clove" to "clove",
        "cans" to "can", "can" to "can",
        "pinches" to "pinch", "pinch" to "pinch",
        "dashes" to "dash", "dash" to "dash",
        "handfuls" to "handful", "handful" to "handful",
        "slices" to "slice", "slice" to "slice",
        "pieces" to "piece", "piece" to "piece",
        "sticks" to "stick", "stick" to "stick",
        "sprigs" to "sprig", "sprig" to "sprig",
        "bunches" to "bunch", "bunch" to "bunch",
        "heads" to "head", "head" to "head",
        "packages" to "package", "package" to "package", "packets" to "packet", "packet" to "packet",
        "jars" to "jar", "jar" to "jar", "bottles" to "bottle", "bottle" to "bottle",
        "boxes" to "box", "box" to "box", "bags" to "bag", "bag" to "bag"
    )

    private val numberWords = mapOf(
        "zero" to "0", "one" to "1", "two" to "2", "three" to "3", "four" to "4",
        "five" to "5", "six" to "6", "seven" to "7", "eight" to "8", "nine" to "9",
        "ten" to "10", "eleven" to "11", "twelve" to "12", "thirteen" to "13",
        "fourteen" to "14", "fifteen" to "15", "sixteen" to "16", "seventeen" to "17",
        "eighteen" to "18", "nineteen" to "19", "twenty" to "20",
        "couple" to "2", "dozen" to "12", "half" to "1/2", "quarter" to "1/4",
        "a" to "1", "an" to "1"
    )


    private val measurementEvidence = Regex(
        "(?i)\\b(?:\\d+(?:\\.\\d+)?|one|two|three|four|five|six|seven|eight|nine|ten|half|quarter|a|an)\\s+(?:tablespoon(?:ful)?s?|teaspoon(?:ful)?s?|tbsp|tsp|cups?|chunks?|grams?|kilograms?|milligrams?|ounces?|pounds?|lbs?|cloves?|cans?|pinches?|dashes?|handfuls?|slices?|pieces?|sticks?|sprigs?|bunches?|heads?|packages?|packets?|jars?|bottles?|boxes?|bags?)\\b"
    )

    fun containsMeasurementEvidence(raw: String): Boolean =
        measurementEvidence.containsMatchIn(normalizeSpeechText(raw))

    fun measurementEvidenceCount(raw: String): Int =
        measurementEvidence.findAll(normalizeSpeechText(raw)).count()

    /** Normalizes speech-recognition variants without changing recipe meaning. */
    fun normalizeSpeechText(raw: String): String {
        return raw
            .replace('’', '\'')
            .replace("½", " 1/2 ")
            .replace("¼", " 1/4 ")
            .replace("¾", " 3/4 ")
            .replace("⅓", " 1/3 ")
            .replace("⅔", " 2/3 ")
            .replace(Regex("(?i)\\b1?\\s*⁄\\s*2\\b"), " 1/2 ")
            .replace(Regex("(?i)\\btea[ -]?spoon(?:ful)?s?\\b"), "teaspoon")
            .replace(Regex("(?i)\\btable[ -]?spoon(?:ful)?s?\\b"), "tablespoon")
            .replace(Regex("(?i)\\bt[ .-]?spoons?\\b"), "teaspoon")
            .replace(Regex("(?i)\\btb[ .-]?spoons?\\b"), "tablespoon")
            .replace(Regex("(?i)\\bfl[ .-]?ounces?\\b"), "fluid ounce")
            .replace(Regex("(?i)\\bcup\\s*ful(?:l)?s?\\b"), "cup")
            .replace(Regex("(?i)\\bcup\\s+fulls?\\b"), "cup")
            .replace(Regex("(?i)\\ba\\s+(half|quarter)\\s+a\\b")) { "${it.groupValues[1]} a" }
            .replace(Regex("(?i)\\b(tbsp|tsp|tablespoons?|teaspoons?)\\s+spoons?\\b"), "$1")
            // Number homophones are corrected only when directly before a measurement unit.
            .replace(Regex("(?i)\\b(?:to|too)\\s+(?=(?:tablespoons?|teaspoons?|tbsp|tsp|cups?|chunks?|grams?|kilograms?|ounces?|pounds?|lbs?|cloves?|cans?|pinches?|slices?|pieces?)\\b)"), "2 ")
            .replace(Regex("(?i)\\bwon\\s+(?=(?:tablespoons?|teaspoons?|cups?|chunks?|grams?|ounces?|pounds?|cloves?|cans?|pinches?)\\b)"), "1 ")
            .replace(Regex("\\s+"), " ")
            .trim()
    }

    fun parse(raw: String): Ingredient {
        val cleaned = normalizeSpeechText(raw)
            .lowercase()
            .trim(' ', ',', '.', ';', ':')
            .replace(Regex("\\s+"), " ")
        if (cleaned.isBlank()) return Ingredient()

        val tokens = cleaned.split(" ").toMutableList()
        val quantityResult = parseQuantity(tokens)
        val quantity = quantityResult.first
        repeat(quantityResult.second) { if (tokens.isNotEmpty()) tokens.removeAt(0) }

        var unit = ""
        if (tokens.isNotEmpty()) {
            val twoWord = tokens.take(2).joinToString(" ").trim(',', '.')
            val oneWord = tokens.first().trim(',', '.')
            when {
                tokens.size >= 2 && unitAliases.containsKey(twoWord) -> {
                    unit = unitAliases.getValue(twoWord)
                    repeat(2) { tokens.removeAt(0) }
                }
                unitAliases.containsKey(oneWord) -> {
                    unit = unitAliases.getValue(oneWord)
                    tokens.removeAt(0)
                }
            }
        }

        var name = tokens.joinToString(" ")
            .replace(Regex("^(?:of\\s+)+(?:the\\s+)?"), "")
            .trim(' ', ',', '.', ';', ':')

        // Remove standalone preparation items that are not ingredients.
        if (Regex("(?i)^\\d+\\s+patties?$" ).matches(name) || Regex("(?i)^patties?$" ).matches(name)) {
            name = ""
        }

        // Remove stray measurement words accidentally captured in the name.
        name = name
            .replace(Regex("(?i)\\b(?:teaspoons?|tsp|tablespoons?|tbsp)\\b"), "")
            .replace(Regex("\\s+"), " ")
            .trim()
            .replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() }

        return Ingredient(quantity = quantity, unit = unit, name = name)
    }

    private fun parseQuantity(tokens: List<String>): Pair<String, Int> {
        if (tokens.isEmpty()) return "" to 0
        val first = tokens[0].trim(',', '.')

        // Common spoken fraction forms: "half a cup", "quarter of a cup".
        if (first in setOf("half", "quarter") && tokens.size >= 2) {
            if (tokens[1] == "a") return (if (first == "half") "1/2" else "1/4") to 2
            if (tokens.size >= 3 && tokens[1] == "of" && tokens[2] == "a") {
                return (if (first == "half") "1/2" else "1/4") to 3
            }
        }

        // Numeric forms: 2, 2.5, 1/2, 2 1/2.
        if (first.matches(Regex("\\d+(?:\\.\\d+)?|\\d+/\\d+"))) {
            if (tokens.size >= 2 && tokens[1].matches(Regex("\\d+/\\d+"))) {
                return "$first ${tokens[1]}" to 2
            }
            if (tokens.size >= 4 && tokens[1] == "and" && tokens[2] in setOf("a", "one") && tokens[3] in setOf("half", "quarter")) {
                return "$first ${if (tokens[3] == "half") "1/2" else "1/4"}" to 4
            }
            if (tokens.size >= 3 && tokens[1] == "and" && tokens[2] in setOf("half", "quarter")) {
                return "$first ${if (tokens[2] == "half") "1/2" else "1/4"}" to 3
            }
            return first to 1
        }

        // "three quarters" / "two thirds".
        if (tokens.size >= 2) {
            val numerator = numberWords[first]?.toIntOrNull()
            val fractionWord = tokens[1].trim(',', '.')
            val denominator = when (fractionWord) {
                "halves", "half" -> 2
                "third", "thirds" -> 3
                "quarter", "quarters", "fourth", "fourths" -> 4
                else -> null
            }
            if (numerator != null && denominator != null && numerator in 1 until denominator) {
                return "$numerator/$denominator" to 2
            }
        }

        val base = numberWords[first] ?: return "" to 0
        if (tokens.size >= 4 && tokens[1] == "and" && tokens[2] in setOf("a", "one") && tokens[3] in setOf("half", "quarter")) {
            return "$base ${if (tokens[3] == "half") "1/2" else "1/4"}" to 4
        }
        if (tokens.size >= 3 && tokens[1] == "and" && tokens[2] in setOf("half", "quarter")) {
            return "$base ${if (tokens[2] == "half") "1/2" else "1/4"}" to 3
        }
        return base to 1
    }
}

File: web/js/ingredient-parser.js
const unitAliases = new Map([
  ['tablespoonfuls','tbsp'], ['tablespoonful','tbsp'], ['tablespoons','tbsp'], ['tablespoon','tbsp'], ['tbsp','tbsp'], ['tbs','tbsp'],
  ['teaspoonfuls','tsp'], ['teaspoonful','tsp'], ['teaspoons','tsp'], ['teaspoon','tsp'], ['tsp','tsp'],
  ['kilograms','kg'], ['kilogram','kg'], ['kg','kg'], ['milligrams','mg'], ['milligram','mg'], ['mg','mg'],
  ['grams','g'], ['gram','g'], ['g','g'], ['liters','L'], ['liter','L'], ['litres','L'], ['litre','L'], ['l','L'],
  ['milliliters','ml'], ['milliliter','ml'], ['millilitres','ml'], ['millilitre','ml'], ['ml','ml'],
  ['fluid ounces','fl oz'], ['fluid ounce','fl oz'], ['fl oz','fl oz'], ['ounces','oz'], ['ounce','oz'], ['oz','oz'],
  ['pounds','lb'], ['pound','lb'], ['lbs','lb'], ['lb','lb'],
  ['cupfuls','cup'], ['cupful','cup'], ['cups','cup'], ['cup','cup'],
  ['chunks','chunk'], ['chunk','chunk'],
  ['cloves','clove'], ['clove','clove'], ['cans','can'], ['can','can'], ['pinches','pinch'], ['pinch','pinch'],
  ['dashes','dash'], ['dash','dash'], ['handfuls','handful'], ['handful','handful'], ['slices','slice'], ['slice','slice'],
  ['pieces','piece'], ['piece','piece'], ['sticks','stick'], ['stick','stick'], ['sprigs','sprig'], ['sprig','sprig'],
  ['bunches','bunch'], ['bunch','bunch'], ['heads','head'], ['head','head'], ['packages','package'], ['package','package'],
  ['packets','packet'], ['packet','packet'], ['jars','jar'], ['jar','jar'], ['bottles','bottle'], ['bottle','bottle'],
  ['boxes','box'], ['box','box'], ['bags','bag'], ['bag','bag']
]);

const numberWords = new Map([
  ['zero','0'], ['one','1'], ['two','2'], ['three','3'], ['four','4'], ['five','5'], ['six','6'], ['seven','7'],
  ['eight','8'], ['nine','9'], ['ten','10'], ['eleven','11'], ['twelve','12'], ['thirteen','13'], ['fourteen','14'],
  ['fifteen','15'], ['sixteen','16'], ['seventeen','17'], ['eighteen','18'], ['nineteen','19'], ['twenty','20'],
  ['couple','2'], ['dozen','12'], ['half','1/2'], ['quarter','1/4'], ['a','1'], ['an','1']
]);

const measurementEvidence = /\b(?:\d+(?:\.\d+)?|one|two|three|four|five|six|seven|eight|nine|ten|half|quarter|a|an)\s+(?:tablespoon(?:ful)?s?|teaspoon(?:ful)?s?|tbsp|tsp|cups?|chunks?|grams?|kilograms?|milligrams?|ounces?|pounds?|lbs?|cloves?|cans?|pinches?|dashes?|handfuls?|slices?|pieces?|sticks?|sprigs?|bunches?|heads?|packages?|packets?|jars?|bottles?|boxes?|bags?)\b/gi;

export function normalizeSpeechText(raw = '') {
  return String(raw)
    .replaceAll('’', "'")
    .replaceAll('½', ' 1/2 ')
    .replaceAll('¼', ' 1/4 ')
    .replaceAll('¾', ' 3/4 ')
    .replaceAll('⅓', ' 1/3 ')
    .replaceAll('⅔', ' 2/3 ')
    .replace(/\b1?\s*⁄\s*2\b/gi, ' 1/2 ')
    .replace(/\btea[ -]?spoon(?:ful)?s?\b/gi, 'teaspoon')
    .replace(/\btable[ -]?spoon(?:ful)?s?\b/gi, 'tablespoon')
    .replace(/\bt[ .-]?spoons?\b/gi, 'teaspoon')
    .replace(/\btb[ .-]?spoons?\b/gi, 'tablespoon')
    .replace(/\bfl[ .-]?ounces?\b/gi, 'fluid ounce')
    .replace(/\bcup\s*ful(?:l)?s?\b/gi, 'cup')
    .replace(/\bcup\s+fulls?\b/gi, 'cup')
    .replace(/\ba\s+(half|quarter)\s+a\b/gi, (_, word) => `${word} a`)
    .replace(/\b(tbsp|tsp|tablespoons?|teaspoons?)\s+spoons?\b/gi, '$1')
    .replace(/\b(?:to|too)\s+(?=(?:tablespoons?|teaspoons?|tbsp|tsp|cups?|chunks?|grams?|kilograms?|ounces?|pounds?|lbs?|cloves?|cans?|pinches?|slices?|pieces?)\b)/gi, '2 ')
    .replace(/\bwon\s+(?=(?:tablespoons?|teaspoons?|cups?|chunks?|grams?|ounces?|pounds?|cloves?|cans?|pinches?)\b)/gi, '1 ')
    .replace(/\s+/g, ' ')
    .trim();
}

export function containsMeasurementEvidence(raw = '') {
  measurementEvidence.lastIndex = 0;
  return measurementEvidence.test(normalizeSpeechText(raw));
}

export function measurementEvidenceCount(raw = '') {
  const text = normalizeSpeechText(raw);
  return [...text.matchAll(new RegExp(measurementEvidence.source, 'gi'))].length;
}

function parseQuantity(tokens) {
  if (!tokens.length) return ['', 0];
  const first = tokens[0].replace(/[,.]+$/g, '');

  if (['half','quarter'].includes(first) && tokens.length >= 2) {
    if (tokens[1] === 'a') return [first === 'half' ? '1/2' : '1/4', 2];
    if (tokens.length >= 3 && tokens[1] === 'of' && tokens[2] === 'a') return [first === 'half' ? '1/2' : '1/4', 3];
  }
  if (first === 'a' && tokens.length >= 2 && ['half','quarter'].includes(tokens[1])) {
    return [tokens[1] === 'half' ? '1/2' : '1/4', 2];
  }

  if (/^(?:\d+(?:\.\d+)?|\d+\/\d+)$/.test(first)) {
    if (tokens.length >= 2 && /^\d+\/\d+$/.test(tokens[1])) return [`${first} ${tokens[1]}`, 2];
    if (tokens.length >= 4 && tokens[1] === 'and' && ['a','one'].includes(tokens[2]) && ['half','quarter'].includes(tokens[3])) {
      return [`${first} ${tokens[3] === 'half' ? '1/2' : '1/4'}`, 4];
    }
    if (tokens.length >= 3 && tokens[1] === 'and' && ['half','quarter'].includes(tokens[2])) {
      return [`${first} ${tokens[2] === 'half' ? '1/2' : '1/4'}`, 3];
    }
    return [first, 1];
  }

  if (tokens.length >= 2) {
    const numerator = Number.parseInt(numberWords.get(first), 10);
    const fractionWord = tokens[1].replace(/[,.]+$/g, '');
    const denominator = ({half:2, halves:2, third:3, thirds:3, quarter:4, quarters:4, fourth:4, fourths:4})[fractionWord];
    if (Number.isFinite(numerator) && denominator && numerator >= 1 && numerator < denominator) return [`${numerator}/${denominator}`, 2];
  }

  const base = numberWords.get(first);
  if (!base) return ['', 0];
  if (tokens.length >= 4 && tokens[1] === 'and' && ['a','one'].includes(tokens[2]) && ['half','quarter'].includes(tokens[3])) {
    return [`${base} ${tokens[3] === 'half' ? '1/2' : '1/4'}`, 4];
  }
  if (tokens.length >= 3 && tokens[1] === 'and' && ['half','quarter'].includes(tokens[2])) {
    return [`${base} ${tokens[2] === 'half' ? '1/2' : '1/4'}`, 3];
  }
  return [base, 1];
}

export function parseIngredient(raw = '') {
  const cleaned = normalizeSpeechText(raw).toLowerCase().trim().replace(/^[\s,.;:]+|[\s,.;:]+$/g, '').replace(/\s+/g, ' ');
  if (!cleaned) return { id: crypto.randomUUID(), quantity:'', unit:'', name:'' };

  const tokens = cleaned.split(' ');
  const [quantity, consumed] = parseQuantity(tokens);
  tokens.splice(0, consumed);

  let unit = '';
  if (tokens.length) {
    const twoWord = tokens.slice(0, 2).join(' ').replace(/[,.]+$/g, '');
    const oneWord = tokens[0].replace(/[,.]+$/g, '');
    if (tokens.length >= 2 && unitAliases.has(twoWord)) {
      unit = unitAliases.get(twoWord);
      tokens.splice(0, 2);
    } else if (unitAliases.has(oneWord)) {
      unit = unitAliases.get(oneWord);
      tokens.splice(0, 1);
    }
  }

  let nameRaw = tokens.join(' ').replace(/^(?:of\s+)+(?:the\s+)?/i, '').replace(/^[\s,.;:]+|[\s,.;:]+$/g, '');

  // Remove standalone preparation items that are not ingredients.
  if (/^\d+\s+patties?$/i.test(nameRaw) || /^patties?$/i.test(nameRaw)) nameRaw = '';

  // Remove stray measurement words accidentally captured in the name.
  nameRaw = nameRaw
    .replace(/\b(?:teaspoons?|tsp|tablespoons?|tbsp)\b/gi, '')
    .replace(/\s+/g, ' ')
    .trim();

  const name = nameRaw ? nameRaw[0].toUpperCase() + nameRaw.slice(1) : '';
  return { id: crypto.randomUUID(), quantity, unit, name };
}

export function displayIngredient(i) {
  return [i.quantity, i.unit, i.name].filter(Boolean).join(' ');
}

export const ingredientParserInternals = { unitAliases, numberWords };
